database.py

```diff
@@ -0,0 +1,125 @@
+import sqlite3
+from elo import ELO
+
+
+class Database:
+    """
+    Class responsible for handle database operations.
+    """
+    def __init__(self):
+        self.conn = sqlite3.connect('./history/cn.db')
+
+    @property
+    def getLastGameID(self):
+        """
+        This method returns id of last saved game
+        @return: integer value of ID
+        """
+        cursor = self.conn.execute("SELECT ID from GAMES ORDER BY ID DESC LIMIT 1")
+        row = cursor.fetchone()
+        if row:
+            return row[0]
+        return -1
+
+    def botExist(self, name):
+        """
+        Method check if bot with this name exist inside database
+        @param name: Name of bot to check existance
+        @return: True or False
+        """
+        cursor = self.conn.execute(f"SELECT ID from BOTS WHERE NAME = '{name}'")
+        row = cursor.fetchone()
+        if row:
+            return True
+        return False
+
+    def insertBot(self, name):
+        """
+        Method used to insert new bot to BOTS table in database
+        @param name: Name of bot
+        @return: integer value of new bot ID
+        """
+        if not self.botExist(name):
+            cursor = self.conn.execute(f"INSERT INTO BOTS (NAME) VALUES('{name}')")
+            self.conn.commit()
+            return cursor.lastrowid
+        return -1
+
+    def getBot(self, name):
+        """
+        Method used to get bot data frmo database
+        @param name: Name of bot inside BOTS table
+        @return: ID, WON, LOST, ELO variables from BOTS table for bot with @name
+        """
+        cursor = self.conn.execute(f"SELECT ID, WON, LOST, ELO from BOTS WHERE NAME = '{name}'")
+        row = cursor.fetchone()
+        if row:
+            return row[0], row[1], row[2], row[3]
+        return -1, -1, -1
+
+    def updateChallangers(self, winner, loser):
+        """
+        Methos used to update Win/Lost/Elo values of bots inside database
+        @param winner: Name of winner bot
+        @param loser: Name of loser bot
+        """
+        winner = self.getBot(winner)
+        loser = self.getBot(loser)
+        winner_elo = winner[3]
+        loser_elo = loser[3]
+        winner_elo, loser_elo = ELO.update_elo(winner_elo, loser_elo)
+        self.conn.execute(f"UPDATE BOTS set WON = {winner[1] + 1}, ELO = {winner_elo} WHERE ID = {winner[0]}")
+        self.conn.execute(f"UPDATE BOTS set LOST = {loser[2] + 1}, ELO = {loser_elo} WHERE ID = {loser[0]}")
```

Comment:
Well, that's not the best idea.
We should not store number of won, lost games in separated column like you did here.
This may cause errors when bot play two games at the same time.
Moreover we dont need to store data that we can easy calculate from GAMES table.

Reply:
In my opinion, the calculation of wins/losses is a much less cumbersome matter than the calculation of ELO values with parallel fights - and yet this is the one you have reservations about. In fact calculating ELO correctly in a situation where the same bot takes part in two games at the same time is in my opinion impossible :D

Another thing is that there is already a lot of code written, which calculates values that can be easily determined from the GAMES table :P

Reply:
Hmm, you are right.
The problem with updating the record is that we lose history and in the event of an error we are unable to recreate and verify the correctness of the ELO.
It seems to me like a situation where the Event Sourcing approach would be better, because by logging each game with information about player ELO and who won, we would be able to recreate the entire history of the ranking.
The question, however, is whether this is a problem we want to focus on or whether we will accept the simplification proposed by you.

For now let's take the easy way 😄

Reply:
Event Sourcing approach is possible with current solution. With use of GAMES table it is possible to recreate ELOs of all bots from BOTS table.